streaming/base/shared.py

```diff
@@ -0,0 +1,149 @@
+"""
+
+import os
+from multiprocessing.shared_memory import SharedMemory
+from time import sleep
+
+import numpy as np
+from filelock import FileLock
+
+TICK = 0.07
+
+
+class SharedBarrier:
+    """A barrier that works inter-process using a file lock and shared memory.
+
+    We set the number of processes (and thereby initialilze num_exit) on the first time this object
+    is called. This is because the object is created in a per-rank process, and called by worker
+    processes.
+
+    Args:
+        filelock_path (str): Path to lock file on local filesystem.
+        shm_path (str): Shared memory object name in /dev/shm.
+    """
+
+    def __init__(self, filelock_path: str, shm_path: str) -> None:
+        self.filelock_path = filelock_path
+        self.shm_path = shm_path
+
+        # Create filelock.
+        dirname = os.path.dirname(filelock_path)
+        os.makedirs(dirname, exist_ok=True)
+        self.lock = FileLock(filelock_path)
+
+        # Create three int32 fields in shared memory: num_enter, num_exit, flag.
+        size = 3 * np.int32(0).nbytes
+        try:
+            self._shm = SharedMemory(shm_path, True, size)
+        except FileExistsError:
+            self._shm = SharedMemory(shm_path, False, size)
+        self._arr = np.ndarray(3, buffer=self._shm.buf, dtype=np.int32)
+        self.num_enter = 0
+        self.num_exit = -1
+        self.flag = True
+
+    def __del__(self):
+        """Destructor clears array that references shm."""
+        try:
+            del self._arr
+        except:
+            pass
+
+    @property
+    def num_enter(self) -> int:
+        """Get property num_enter.
+
+        Returns:
+            int: Number of processes that have entered the barrier.
+        """
+        return self._arr[0]
+
+    @num_enter.setter
+    def num_enter(self, num_enter: int) -> None:
+        """Set property num_enter.
+
+        Args:
+            num_enter (int): Number of processes that have entered the barrier.
+        """
+        self._arr[0] = num_enter
+
+    @property
+    def num_exit(self) -> int:
+        """Get property num_exit.
+
+        Returns:
+            int: Number of processes that have exited the barrier.
+        """
+        return self._arr[1]
+
+    @num_exit.setter
+    def num_exit(self, num_exit: int) -> None:
+        """Set property num_exit.
+
+        Args:
+            num_exit (int): Number of processes that have exited the barrier.
+        """
+        self._arr[1] = num_exit
+
+    @property
+    def flag(self) -> bool:
+        """Get property flag.
+
+        Returns:
+            bool: The flag value.
+        """
+        return bool(self._arr[2])
+
+    @flag.setter
+    def flag(self, flag: bool) -> None:
+        """Set property flag.
+
+        Args:
+            flag (bool): The flag value.
+        """
+        self._arr[2] = bool(flag)
+
+    def __call__(self, num_procs: int) -> None:
```

Comment:
Why pass the `num_procs` here? Can we pass the `num_procs` in the `__init__()` method of `class SharedBarrier` ? The shared barrier object will be initialized only for one cluster configuration and `world.num_workers` will not going to change in one cluster configuration run.

Reply:
we don't know `num_workers` when per-rank copy of streaming dataset is initialiized